### PR TITLE
[Raiden Weight Sync 2/6] FFI transport selection, preflight mismatch check, and host-stage normalization

### PR DESCRIPTION
**Pull request by YixuanWang-99**
## Overview
Part of the stacked Raiden weight-sync enablement PR chain replacing #2083.
**Pairs with MaxText [M4]** (`AI-Hypercomputer/maxtext`).

**Stack:**
- [T1] https://github.com/google/tunix/pull/2146: Multi-host discovery & multi-axis sharding in RaidenHandler
- **[T2a] google/tunix (this PR)**: Transport selection, FFI preflight mismatch check, and host-stage normalization
- [T3] google/tunix: Extract RaidenDestinationWeightSyncMixin
- [T4] google/tunix: Cache lifecycle and prefix caching
- [T5] google/tunix: MoE 128-lane interleaving for TPU GMM layout (pairs with MaxText M2)
- [T6] google/tunix: MaxText trainer configuration plumbing

## Details
- **The Step 0 Fix**: Solves the end-to-end FFI path failure by centralizing transport resolution.
- Adds and exports `resolve_use_ffi(explicit: bool | str | None, *, is_proxy: bool | None) -> bool` and `normalize_host_stage(host_stage: bool | None, *, use_ffi: bool, is_proxy: bool) -> bool`.
- `WeightSyncCoordinator` adds preflight validation rejecting any pairing of mismatched source and destination transport modes before scheduling any transfers.
- `WorkUnitMetadata` carries resolved `transport_mode` ("ffi" / "tcp") and `use_ffi`.
- Integrates `release_host_arrays()` and checksum / tensor element count diagnostics.

## Verification
- `pytest tests/experimental/weight_sync/raiden_synchronizer_test.py` (32/32 passed)
- `pytest tests/experimental/weight_sync/weight_sync_coordinator_test.py` (73/73 passed)

### CODE REVIEW

File: tunix/experimental/weight_sync/raiden_synchronizer.py [hunk 1]
```diff
@@ -48,25 +50,44 @@ def _log_rss(tag: str) -> None:
...
+  if name == "_raiden_ffi":
+    return _get_raiden_ffi()
+  if name == "_ws_lib":
+    return _get_ws_lib()
+  raise AttributeError(f"module {__name__!r} has no attribute {name!r}")
 
 
 def _ensure_ffi_compute_on_compat() -> None:
   """Bridges TPU-sync wheels that call the newer compute_on decorator API."""
-  try:
```
Comment (wang2yn84) [inline tunix/experimental/weight_sync/raiden_synchronizer.py]: This is necessary, according to Raiden, it's tested based on Jax 0.11.0, and the compute_on API has breaking change from 0.11.0 and 0.11.1. @datenglin to confirm.

File: tunix/experimental/weight_sync/raiden_synchronizer.py [hunk 2]
```diff
@@ -124,6 +156,32 @@ def unpack_ip(row: Any) -> str:
   return f"[{addr_str}]" if ":" in addr_str else addr_str
 
 
+def to_host_cpu_state(state: Any) -> Any:
```
Comment (wang2yn84) [inline tunix/experimental/weight_sync/raiden_synchronizer.py]: Remove it, we should only use FFI for Pathways and never stage on src host.

File: tunix/experimental/weight_sync/weight_sync_coordinator.py [hunk 3]
```diff
@@ -165,7 +166,8 @@ def create_default_handler(
     from tunix.experimental.weight_sync import raiden_handler
 
     handler = raiden_handler.RaidenHandler(
-        transfer_options=raiden_handler.make_host_staged_transfer_options()
+        name_resolver=name_resolver,
```
Comment (wang2yn84) [inline tunix/experimental/weight_sync/weight_sync_coordinator.py]: Is it used by another PR?

File: tunix/experimental/weight_sync/weight_sync_coordinator.py [hunk 4]
```diff
@@ -1009,6 +1012,35 @@ async def record_workers(final_error: str = "") -> None:
...
+              side,
+              m.unit.job_name,
+              m.unit.job_replica_id,
+              len(m.shards),
+              list(m.shards),
+          )
+      src_shards = sum(len(m.shards) for m in src_metadata)
+      for m in dst_metadata:
+        if src_shards and len(m.shards) != src_shards:
+          logging.warning(
```
Comment (wang2yn84) [inline tunix/experimental/weight_sync/weight_sync_coordinator.py]: Is this for debugging only? Source shards != destination shards should be normal?

File: tunix/experimental/weight_sync/weight_sync.py [hunk 5]
```diff
@@ -76,9 +79,13 @@ class TensorMetadata:
     layout: Layout mapping.
     item_size: Bytes per element.
     layer_idx: Stable batching ordinal.
-    sharding_spec: One mesh axis name per TENSOR dimension, empty string where
-      that dimension is replicated. This is the subset of JAX `PartitionSpec`
-      used by the Tunix/JAX adapters: `P(None, "y")` is `("", "y")`. Together
+    sharding_spec: The mesh axis name sharding each TENSOR dimension, empty
```
Comment (wang2yn84) [inline tunix/experimental/weight_sync/weight_sync.py:82]: This may not be the ideal protocol to use a string when it becomes complicated, a more structured tensor_dim_axes = (
    (),                    # replicated
    ("tp",),               # single axis
    ("attention_dp", "tp") # multi-axis on one tensor dim
) might be better, but the parsing will also need to change. If it's too much change, please at least add a TODO.

File: tunix/experimental/weight_sync/weight_sync.py [hunk 6]
```diff
@@ -186,6 +194,8 @@ class WorkUnitMetadata:
   item_size: Optional[int] = None
   variables: tuple[TensorMetadata, ...] = ()
   mesh_axes: Optional[tuple[str, ...]] = None
+  transport_mode: Optional[str] = None
```
Comment (wang2yn84) [inline tunix/experimental/weight_sync/weight_sync.py:206]: Please update doc string.

File: tunix/experimental/weight_sync/raiden_synchronizer.py [hunk 7]
```diff
@@ -197,14 +276,74 @@ def _axis_name(axis: Any) -> str:
...
+  fills only its own share, leaving the rest of its SetGlobalShardIndices at
+  -1 -- the transfer then completes green while delivering only the shards one
+  host happened to own.
+
+  `process_index` alone is wrong under Pathways: the client is a single process
+  driving every worker, so all proxy devices report 0 and this collapses to
+  len(devices). Prefer whichever attribute actually distinguishes the workers,
+  and fall back to the per-host hardware ordinal.
+  """
+  env = os.environ.get("RAIDEN_DEVICES_PER_HOST")
```
Comment (wang2yn84) [inline tunix/experimental/weight_sync/raiden_synchronizer.py:276]: Why do we need this? Does raiden have different view of devices per host than Jax?

File: tunix/experimental/weight_sync/raiden_synchronizer.py [hunk 8]
```diff
@@ -197,14 +276,74 @@ def _axis_name(axis: Any) -> str:
...
+    if len(groups) > 1 and len(devices) % len(groups) == 0:
+      return len(devices) // len(groups)
+  local_ids = {getattr(d, "local_hardware_id", None) for d in devices}
+  local_ids.discard(None)
+  if 1 < len(local_ids) < len(devices) and len(devices) % len(local_ids) == 0:
+    return len(local_ids)
+  return len(devices)
+
+
+def _reduce_mesh(mesh: Any) -> Any:
```
Comment (wang2yn84) [inline tunix/experimental/weight_sync/raiden_synchronizer.py]: Do we still need this? b/557061810 might have fixed it?  @datenglin  to confirm.

File: tunix/experimental/weight_sync/raiden_synchronizer.py [hunk 9]
```diff
@@ -217,14 +356,79 @@ def _tensor_metadata(name: str, arr: Any, layer_idx: int):
   )
 
 
+def resolve_use_ffi(
```
Comment (wang2yn84) [inline tunix/experimental/weight_sync/raiden_synchronizer.py]: I don't think we need such complicated logic, for Pathways, we should always use FFI before the PJRT path is available.

File: tunix/experimental/weight_sync/raiden_synchronizer.py [hunk 10]
```diff
@@ -217,14 +356,79 @@ def _tensor_metadata(name: str, arr: Any, layer_idx: int):
...
+      # Default under Pathways proxy is FFI on; non-proxy is FFI off.
+      desired = is_proxy
+
+  if not desired:
+    return False
+
+  return _get_raiden_ffi() is not None
+
+
+def normalize_host_stage(
```
Comment (wang2yn84) [inline tunix/experimental/weight_sync/raiden_synchronizer.py]: Remove, we don't need host stage at all.

File: tunix/experimental/weight_sync/raiden_synchronizer.py [hunk 11]
```diff
@@ -234,21 +438,30 @@ def __init__(
       *,
       worker_index: int = 0,
       auto_h2d: bool = False,
+      host_stage: Optional[bool] = None,
+      use_ffi: Optional[bool] = None,
```
Comment (wang2yn84) [inline tunix/experimental/weight_sync/raiden_synchronizer.py]: I challenge if we need use_ffi at all. Cuz that's the only option for Pathways.

File: tunix/experimental/weight_sync/raiden_synchronizer.py [hunk 12]
```diff
@@ -341,19 +578,6 @@ def _init_ffi_transport(self, *, is_d2h: bool) -> None:
           devices_per_host,
       )
 
-      # TODO(b/557061810): Re-enable this once the bug is fixed and the FFI
```
Comment (wang2yn84) [inline tunix/experimental/weight_sync/raiden_synchronizer.py:344]: Let's keep it.

File: tunix/experimental/weight_sync/raiden_synchronizer.py [hunk 13]
```diff
@@ -513,13 +757,127 @@ def d2h(self) -> None:
...
-    if self._is_proxy:
+    if self._use_ffi:
       self._ffi_h2d()
       return
     if self._sync is not None:
       self._sync.h2d()
       jax.block_until_ready(self.arrays)
 
+  def apply_to_runner(self, runner: Any) -> None:
+    """Applies updated arrays after H2D to the runner's state_leaves and state."""
```
Comment (wang2yn84) [inline tunix/experimental/weight_sync/raiden_synchronizer.py:672]: When we bind the runner's state, h2d indeed means to apply to it, isn't it? Why do we need this?

File: tunix/experimental/weight_sync/raiden_synchronizer.py [hunk 14]
```diff
@@ -513,13 +757,127 @@ def d2h(self) -> None:
...
+        jax.tree_util.tree_structure(runner.state), new_leaves
+    )
+    logging.info(
+        "%s apply_to_runner: successfully applied %d arrays to runner state and state_leaves (total runner leaves: %d).",
+        self.job_name,
+        len(matched_indices),
+        len(new_leaves),
+    )
+
+  def release_host_arrays(self) -> None:
```
Comment (wang2yn84) [inline tunix/experimental/weight_sync/raiden_synchronizer.py]: Delete

File: tunix/experimental/weight_sync/raiden_synchronizer.py (hunk 8 above)
Comment (datenglin) [reply tunix/experimental/weight_sync/raiden_synchronizer.py]: No we do not need this. Yes the issue has been fixed (though let's not mention the bug number here next time :-)).

File: tunix/experimental/weight_sync/raiden_synchronizer.py (hunk 13 above)
Comment (khatwanimohit) [reply tunix/experimental/weight_sync/raiden_synchronizer.py:672]: You're right in principle — _ffi_h2d rebinds in place, so h2d should be enough. It isn't today because runner.state_leaves is a separate tuple that doesn't share identity with our arrays, hence the name matcher. The real fix is to record the runner's leaf identity at bind() so h2d writes back directly, deleting the matcher entirely. 

Bigger than this PR, can do a cleanup in a followup PR

File: tunix/experimental/weight_sync/raiden_synchronizer.py (hunk 8 above)
Comment (khatwanimohit) [reply tunix/experimental/weight_sync/raiden_synchronizer.py]: removed

File: tunix/experimental/weight_sync/raiden_synchronizer.py (hunk 10 above)
Comment (khatwanimohit) [reply tunix/experimental/weight_sync/raiden_synchronizer.py]: removed

File: tunix/experimental/weight_sync/weight_sync.py (hunk 6 above)
Comment (khatwanimohit) [reply tunix/experimental/weight_sync/weight_sync.py:206]: done

File: tunix/experimental/weight_sync/weight_sync.py (hunk 5 above)
Comment (khatwanimohit) [reply tunix/experimental/weight_sync/weight_sync.py:82]: added a TODO

File: tunix/experimental/weight_sync/raiden_synchronizer.py (hunk 2 above)
Comment (khatwanimohit) [reply tunix/experimental/weight_sync/raiden_synchronizer.py]: removed

File: tunix/experimental/weight_sync/raiden_synchronizer.py (hunk 13 above)
Comment (wang2yn84) [reply tunix/experimental/weight_sync/raiden_synchronizer.py]: Sounds good, let's add a TODO here.

File: tunix/experimental/weight_sync/raiden_synchronizer.py [hunk 15]
```diff
@@ -292,8 +404,17 @@ def _init_ffi_transport(self, *, is_d2h: bool) -> None:
...
+    # Mesh POSITION, not device id. The controller indexes a source shard by
+    # its position in the mesh (`_get_global_indices` walks
+    # physical_mesh_shape), while the native layer keys staging off whatever we
+    # pass here -- slot = shard_idx % num_shards, submanager = shard_idx /
+    # num_shards, and SetGlobalShardIndices records it as the global index.
+    # create_device_mesh reorders devices for topology (a 2x2x2 v5p slice comes
+    # back as ids [0,1,3,2,6,7,5,4]), so keying off d.id labels each slice with
+    # the wrong global index. A 2x2x1 slice happens to be identity-ordered,
+    # which is why this only ever showed up multi-host.
+    global_ids = jnp.arange(
```
Comment (wang2yn84) [inline tunix/experimental/weight_sync/raiden_synchronizer.py]: Hmm, from Raiden example, it's like using device id. @datenglin to comment.

File: tunix/experimental/weight_sync/weight_sync_coordinator.py (hunk 3 above)
Comment (khatwanimohit) [reply tunix/experimental/weight_sync/weight_sync_coordinator.py]: removed

File: tunix/experimental/weight_sync/raiden_synchronizer.py (hunk 15 above)
Comment (wang2yn84) [reply tunix/experimental/weight_sync/raiden_synchronizer.py]: Dateng confirmed offline.

**Review by tianshub — APPROVED** [review]
